### PR TITLE
[Merged by Bors] - layerfetcher: fail tx hashes fetch when at least 1 fails

### LINKED ISSUE #3139: hare, proposals: don't vote for proposals with missing transactions
if a proposal with the transaction that is not available to any node gets into hare output it will halt the network. since we don't plan to prune transactions the simplest solution is to change this https://github.com/spacemeshos/go-spacemesh/blob/develop/layerfetcher/layers.go#L674-L688 method so that it will fail even if single transaction is not found.

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

Closes #3139

## Changes

- fail tx hashes fetch when any of them fails
- change atx/tx to share the same code with ballot/proposal/ballots in layerfetcher and add tests

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by countvonzero** [comment]
bors try

**Review by dshulyak — APPROVED** [review]

File: layerfetcher/layers.go [hunk 1]
```diff
@@ -552,27 +542,14 @@ func (l *Logic) FetchAtx(ctx context.Context, id types.ATXID) error {
...
-		}
-		if !res.IsLocal {
-			err := l.getAtxResults(ctx, res.Hash, res.Data)
-			if err != nil {
-				return err
-			}
-		}
+	l.log.WithContext(ctx).With().Debug("requesting atxs from peer", log.Int("num_atxs", len(ids)))
+	hashes := types.ATXIDsToHashes(ids)
+	if errs := l.getHashes(ctx, hashes, fetch.ATXDB, l.atxHandler.HandleAtxData); len(errs) > 0 {
```
Comment (evt) [inline layerfetcher/layers.go:551]: is it a common practice to return just the first error?

**Review by evt — APPROVED** [review]

File: layerfetcher/layers.go (hunk 1 above)
Comment (countvonzero) [reply layerfetcher/layers.go:551]: ppl tend to invent their own solutions with returning multiple errors. some wrap multiple errors with errors.Wrap but i don't really like that, as for me that implies their are in the same chain of event and is confusing. some created their own customized error to include a list of errors. in this case, i just don't care enough to do that.

my rationale here is that when i'm debugging, i'll look at the log anyway and each of the missing hash has its corresponding log. the point here is to fail the upstream Block/Proposal handler. so i took the simplest approach to just return the first error.

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**